File: MicroTDCore/Sources/MicroTDCore/Systems/RenderSnapshot.swift
// RenderSnapshot.swift
// Rendering-only data transfer objects
// SpriteKit consumes these DTOs, NEVER Core entity types

import Foundation

/// Render-only enemy data (id + position)
public struct RenderEnemy {
    public let id: Int
    public let type: String
    public let pathProgress: Double
    
    public init(id: Int, type: String, pathProgress: Double) {
        self.id = id
        self.type = type
        self.pathProgress = pathProgress
    }
}

/// Render-only tower data (id + grid position)
public struct RenderTower {
    public let id: Int
    public let type: String
    public let gridX: Int
    public let gridY: Int
    
    public init(id: Int, type: String, gridX: Int, gridY: Int) {
        self.id = id
        self.type = type
        self.gridX = gridX
        self.gridY = gridY
    }
}

/// Complete render snapshot (no gameplay state, only what Sprite needs to render)
public struct RenderSnapshot {
    public let enemies: [RenderEnemy]
    public let towers: [RenderTower]
    
    public init(enemies: [RenderEnemy], towers: [RenderTower]) {
        self.enemies = enemies
        self.towers = towers
    }
}

File: expo/src/core/systems/RenderSnapshot.ts
// RenderSnapshot.ts
// Read-only DTOs for rendering — ported from MicroTDCore/Systems/RenderSnapshot.swift

import type { GridPosition } from '../entities/GridPosition';

/**
 * Rendering data for a single enemy (no gameplay logic exposed)
 */
export interface RenderEnemy {
    id: number;
    typeID: string;
    pathProgress: number;
    hpFraction: number;  // 0.0 to 1.0
    isBoss: boolean;
    isSlowed: boolean;
}

/**
 * Rendering data for a single tower (no gameplay logic exposed)
 */
export interface RenderTower {
    id: number;
    typeID: string;
    gridPosition: GridPosition;
    kills: number;
    totalDamage: number;
}

/**
 * Complete snapshot of the game state for the rendering layer.
 * Contains only the data needed for visualization — no mutable game logic.
 */
export interface RenderSnapshot {
    tick: number;
    enemies: RenderEnemy[];
    towers: RenderTower[];
    coins: number;
    lives: number;
    currentWave: number;
    totalWaves: number;
    stateType: string;  // RunStateType
}
